File: LISTAS/LISTA ENLAZADA CIRCULAR/LISTACIRCULAR.py
class NodoCircular:
    def __init__(self, dato):
        self.dato = dato
        self.siguiente = None

class ListaCircular:
    def __init__(self):
        self.ultimo = None
    
    def insertar_vacio(self, dato):
        nuevo = NodoCircular(dato)
        nuevo.siguiente = nuevo
        self.ultimo = nuevo
    
    def insertar_inicio(self, dato):
        if not self.ultimo:
            self.insertar_vacio(dato)
            return
        
        nuevo = NodoCircular(dato)
        nuevo.siguiente = self.ultimo.siguiente
        self.ultimo.siguiente = nuevo
    
    def insertar_final(self, dato):
        if not self.ultimo:
            self.insertar_vacio(dato)
            return
        
        nuevo = NodoCircular(dato)
        nuevo.siguiente = self.ultimo.siguiente
        self.ultimo.siguiente = nuevo
        self.ultimo = nuevo
    
    def eliminar(self, dato):
        if not self.ultimo:
            return
        
        # Caso único elemento
        if self.ultimo.siguiente == self.ultimo:
            if self.ultimo.dato == dato:
                self.ultimo = None
            return
        
        actual = self.ultimo.siguiente
        anterior = self.ultimo
        
        while True:
            if actual.dato == dato:
                if actual == self.ultimo:
                    self.ultimo = anterior
                anterior.siguiente = actual.siguiente
                return
            
            anterior = actual
            actual = actual.siguiente
            if actual == self.ultimo.siguiente:
                break
    
    def mostrar(self):
        if not self.ultimo:
            return []
        
        elementos = []
        actual = self.ultimo.siguiente
        
        while True:
            elementos.append(actual.dato)
            actual = actual.siguiente
            if actual == self.ultimo.siguiente:
                break
        
        return elementos

File: LISTAS/LISTA ENLAZADA CIRCULAR/LISTACIRCULAR.js
class NodoCircular {
    constructor(dato) {
        this.dato = dato;
        this.siguiente = null;
    }
}

class ListaCircular {
    constructor() {
        this.ultimo = null;
    }
    
    insertarVacio(dato) {
        const nuevo = new NodoCircular(dato);
        nuevo.siguiente = nuevo;
        this.ultimo = nuevo;
    }
    
    insertarInicio(dato) {
        if (!this.ultimo) {
            this.insertarVacio(dato);
            return;
        }
        
        const nuevo = new NodoCircular(dato);
        nuevo.siguiente = this.ultimo.siguiente;
        this.ultimo.siguiente = nuevo;
    }
    
    insertarFinal(dato) {
        if (!this.ultimo) {
            this.insertarVacio(dato);
            return;
        }
        
        const nuevo = new NodoCircular(dato);
        nuevo.siguiente = this.ultimo.siguiente;
        this.ultimo.siguiente = nuevo;
        this.ultimo = nuevo;
    }
    
    eliminar(dato) {
        if (!this.ultimo) return;
        
        // Caso único elemento
        if (this.ultimo.siguiente === this.ultimo) {
            if (this.ultimo.dato === dato) {
                this.ultimo = null;
            }
            return;
        }
        
        let actual = this.ultimo.siguiente;
        let anterior = this.ultimo;
        
        do {
            if (actual.dato === dato) {
                if (actual === this.ultimo) {
                    this.ultimo = anterior;
                }
                anterior.siguiente = actual.siguiente;
                return;
            }
            
            anterior = actual;
            actual = actual.siguiente;
        } while (actual !== this.ultimo.siguiente);
    }
    
    mostrar() {
        if (!this.ultimo) return [];
        
        const elementos = [];
        let actual = this.ultimo.siguiente;
        
        do {
            elementos.push(actual.dato);
            actual = actual.siguiente;
        } while (actual !== this.ultimo.siguiente);
        
        return elementos;
    }
}